### PR TITLE
feat(browser): playwright provider doesn't allow resizing the browser viewport

### PR DESCRIPTION
**Pull request by userquin**
### Description

Playwright using custom viewport implementation, once the browser is opened PW will prevent resizing the Vitest UI.

We need to configure `viewport: null` when creating the context, for context check: https://github.com/microsoft/playwright/issues/20721

_Playwright prevents resizing the viewport_
imagen

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by userquin** [comment]
@sheremet-va we can also open it maximized via args if you want to add it here: https://discord.com/channels/807756831384403968/1255831070339891281/1255867837793439857

```ts
launchOptions: { args: ["--start-maximized"] }
```

**Comment by kwojcik** [comment]
Is this the same problem as what causes my failure screenshots to get cutoff? This screenshot is taken with `viewport: { width: 800, height: 1500 }`
screenshot

**Comment by sheremet-va** [comment]
> we can also open it maximized via args if you want to add it here

Maybe we should open it in max if `--browser.ui` is true, and pass down the same viewport  from the config otherwise

**Comment by userquin** [comment]
> > we can also open it maximized via args if you want to add it here
> 
> Maybe we should open it in max if `--browser.ui` is true, and pass down the same viewport from the config otherwise

Added `launchViewport` to resolved browser options: Vitest UI maximized when ui enabled with PW and WDIO providers.

For WDIO I haven't found a way to maximize the window on `remove` call, using [browser.maximizeWindow](https://webdriver.io/docs/api/webdriver/#maximizewindow), on the browser opens , the window is maximized.

For PW, adding `--start-maximized` arg to `launch` (when missing `--start-maximized` and `--start-fullscreen` from args) just works: Vitest UI maximized when the browser is opened => https://peter.sh/experiments/chromium-command-line-switches/

**Comment by sheremet-va** [comment]
Let’s not add any new options. Args can already be overriden by providerOptions

File: packages/browser/src/node/providers/playwright.ts [hunk 1]
```diff
@@ -85,11 +95,15 @@ export class PlaywrightBrowserProvider implements BrowserProvider {
...
 
     const browser = await this.openBrowser()
-    const context = await browser.newContext({
+    const options = {
       ...this.options?.context,
       ignoreHTTPSErrors: true,
       serviceWorkers: 'allow',
-    })
+    } satisfies BrowserContextOptions
+    if (this.ctx.config.browser.launchViewport.maximized) {
```
Comment (sheremet-va) [inline packages/browser/src/node/providers/playwright.ts:103]: What I _was_ proposing is using the `browser.viewport` values for non-ui instances

Comment (userquin) [reply packages/browser/src/node/providers/playwright.ts:103]: we cannot use that entry, that viewport is about the iframe not the Vitest UI (top context)

Comment (userquin) [reply packages/browser/src/node/providers/playwright.ts:103]: how can we display Vitest UI in mobile viewport? no makes sense

Comment (sheremet-va) [reply packages/browser/src/node/providers/playwright.ts]: > for non-ui instances

When UI is disabled

Comment (userquin) [reply packages/browser/src/node/providers/playwright.ts]: when ui is disabled this PR doesn't modify the options, and so the options are passed to the provider (I'll check it again, PW for sure in the `newContext` call => `createContext` in playwright.ts module)

Comment (userquin) [reply packages/browser/src/node/providers/playwright.ts]: we need to review the logic to ensure the iframe is "visible", if we use the viewport size in headless mode, the iframe maybe not being "displayed" properly

Comment (userquin) [reply packages/browser/src/node/providers/playwright.ts]: sorry Vlad, forgot it...

Comment (userquin) [reply packages/browser/src/node/providers/playwright.ts]: adding viewport size when ui disabled...

Comment (userquin) [reply packages/browser/src/node/providers/playwright.ts]: too many things :sweat_smile:

Comment (userquin) [reply packages/browser/src/node/providers/playwright.ts]: done, included also a new `viewport.test.ts`

File: test/browser/test/viewport.test.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    const { width: actualWidth, height: actualHeight } = window.document.documentElement.getBoundingClientRect()
+
+    // eslint-disable-next-line no-console
+    console.log({ width, height, actualWidth, actualHeight })
+
+    expect(actualWidth).toBe(width)
+    expect(actualHeight).toBe(height)
+  })
+
+  it.skipIf(page.config.browser.headless)('window has been maximized', async () => {
```
Comment (userquin) [inline test/browser/test/viewport.test.ts:16]: remove async

File: test/browser/test/viewport.test.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { page, server } from '@vitest/browser/context'
+import { describe, expect, it } from 'vitest'
+
+describe.skipIf(server.provider === 'preview')('viewport window has been properly initialized', () => {
+  it.skipIf(!page.config.browser.headless)('viewport has proper width and size', () => {
```
Comment (userquin) [inline test/browser/test/viewport.test.ts:5]: change this test: `'viewport has proper size' or `'viewport has proper width and height'

**Comment by userquin** [comment]
We can use browser capabilities, but will differ between browsers (note from @christian-bromann ):
```ts
capabilities: {
  browserName: 'chrome',
  'goog:chromeOptions': {
    args: ['--start-maximized', '--start-fullscreen']
  }
}
```

File: packages/browser/src/node/providers/playwright.ts (hunk 1 above)
Comment (userquin) [reply packages/browser/src/node/providers/playwright.ts]: tested with WDIO with chrome, edge and firefox with and without headless in my local, cannot test safari...

File: test/browser/test/viewport.test.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import { page, server } from '@vitest/browser/context'
+import { describe, expect, it } from 'vitest'
+
+describe.skipIf(server.provider === 'preview')('viewport window has been properly initialized', () => {
+  it.skipIf(!page.config.browser.headless)('viewport has proper size', () => {
+    const { width, height } = page.config.browser.viewport
+    const { width: actualWidth, height: actualHeight } = window.document.documentElement.getBoundingClientRect()
+
+    // eslint-disable-next-line no-console
+    console.log({ width, height, actualWidth, actualHeight })
```
Comment (sheremet-va) [inline test/browser/test/viewport.test.ts:10]: why do you need logs here?

File: packages/browser/src/node/providers/webdriver.ts [hunk 5]
```diff
@@ -80,6 +80,14 @@ export class WebdriverBrowserProvider implements BrowserProvider {
       capabilities: this.buildCapabilities(),
     })
 
+    if (this.ctx.config.browser.ui) {
+      await this.browser.maximizeWindow()
+    }
+    else {
+      const { width, height } = this.ctx.config.browser.viewport
+      await this.browser.setWindowSize(width, height)
```
Comment (sheremet-va) [inline packages/browser/src/node/providers/webdriver.ts:88]: You do it here by not in playwright, is this intentional? Do you think we should keep viewport only for the iframe?

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/browser/src/node/providers/webdriver.ts (hunk 5 above)
Comment (userquin) [reply packages/browser/src/node/providers/webdriver.ts:88]: There is no need (intentional), PW will use the viewport size properly from the options, WDIO will ignore them.

File: test/browser/test/viewport.test.ts (hunk 4 above)
Comment (userquin) [reply test/browser/test/viewport.test.ts:10]: removing the logs...

Comment (userquin) [reply test/browser/test/viewport.test.ts]: done

File: packages/browser/src/node/providers/webdriver.ts (hunk 5 above)
Comment (userquin) [reply packages/browser/src/node/providers/webdriver.ts:88]: > Do you think we should keep viewport only for the iframe?

Yes, why not?

**Comment by userquin** [comment]
With latest change, the `viewport.test.ts` will not pass when running it with Vitest UI using Firefox or Safari (with both providers).

**Review by sheremet-va — APPROVED** [review]

**Pull request merged**